### ISSUE TITLE
Investigate on dynamic importer handling in Python

### ISSUE THREAD

**Issue by ChristophWWagner** [EMS-TU-Ilmenau/chefkoch#51]
Find out how to compile any python file (given by filename) into a python compiled bytecode container (.pyc) and how to import these .pyc (that may also lay outside a module) into the python namespace as an object.

There shall be functions that
 * Compile a python file
 * Import python scripts and compiled objects
 * check if a file contains a python module or a compiled bytecode object

**Comment by JanKirchhofTU** [comment]
[This](http://opac.lbs-ilmenau.gbv.de/DB=1/XMLPRS=N/PPN?PPN=1691753580) might be a good source to start with.

**Comment by fabiankrieg** [comment]
He links the [following website](https://nedbatchelder.com/blog/200804/the_structure_of_pyc_files.html) in the DA which might be of use.

**Comment by makley273** [comment]
Compiling should be relatively easy with py_compile or compileall

[Methods for shell and in python commands shown here](https://www.csestack.org/convert-py-pyc-python-compile/#:~:text=Method%201%3A%20using%20py_compile%20module,with%20the%20file%20extension%20pyc%20)

[Also here](https://stackoverflow.com/questions/5607283/how-can-i-manually-generate-a-pyc-file-from-a-py-file)

**Comment by makley273** [comment]
Importing pyc files should gereally be possible just like py files, but pyc files are highly specific to OS, python version, architecture, etc.
For importing pyc files from others may decompilers like [uncompyle6](https://github.com/rocky/python-uncompyle6/) or [decompyle3](https://github.com/rocky/python-decompile3) should be used to decompyle pyc and rather import the generated py files

**Comment by ChristophWWagner** [comment]
yes, pyc files are very short-lived, depending mostly on the local version of python. Since in the scope of a chefkoch project, the python version should be the same, this should not be a hindering factor.

However, maybe it is a good idea to also not forget what the original source file was. Just in case we later on might come across a situation where we need to handle special cases from version hell. Good point! I'll take this into consideration

**Comment by pegro** [comment]
May I ask what the use case is here?
The Python interpreter will automatically create those files when a module is loaded. Is this somehow about trying to avoid compiling? That shouldn't be a huge performance hit, since python byte code is not highly optimized.

**Comment by makley273** [comment]
> May I ask what the use case is here?
> The Python interpreter will automatically create those files when a module is loaded. Is this somehow about trying to avoid compiling? That shouldn't be a huge performance hit, since python byte code is not highly optimized.

@ChristophWWagner

**Comment by ChristophWWagner** [comment]
We want to be able to fingerprint python code without beeing too affected by refactoring. Since the project focuses on integrity and reproducibility, we need to consider the source code for each processing step to be volatile as well. Therefore we need to detect whenever changes were made to that code and rebuild everything that depends on it.

From the single-responsibility-principle, it looks like it is the best solution to handle the actual step source just like any other resource and let the `Fridge` do all the integrity and rebuilding decisions based on the resource of that code.

One could just use the .py file here, however since we do not really want to rebuild almost everything when just things like timestamps, code formatting or -comments changes, the idea is to let the `StepPython` class handle the specifics as it is a specialized version of `Item` by handling step scripts as python bytecode files (which do not contain comments or formatting anymore). This should increase robustness against automated text representation modifiers greatly, such as git, black or others.

**Comment by pegro** [comment]
I see.
But the pyc still cotains a timestamp of the source file, so you would need to look at the actual code object stored inside.
Also note that, as you said comments are omitted for pyc files, but dosctrings are still very much part of the code, to allow access to ``__doc__``.

So either you make sure you omit the first 8 bytes (4 bytes python version dependent magic code + 4 bytes timestamp of source code) of a pyc file when hashing/fingerprinting.
Or you find other means to do it.

Also careful when you just look at the bytecode stored in a code object (co_code). It doesn't contain constants or variable names, which might change the behavior of the code drastically.

So I'm not sure of how many files we talk here, but it might be worth looking at the [``ast``](https://docs.python.org/3.8/library/ast.html#module-ast) module, which allows access to the AST after parsing the python code. Based on that you could find a way to detect changes. I found a [discussion on a mailing list](https://mail.python.org/pipermail/python-ideas/2013-September/023475.html) using that approach.
